### ISSUE TITLE
Add ability to tag particles

### ISSUE THREAD

**Issue by jacanchaplais** [jacanchaplais/graphicle#50]
Already can group particles by their parents in a topological sense, using `graphicle.select.hard_descendants()`. At hadronisation vertices, the ancestry becomes mixed, so only using topological information of the generation DAG results particles outgoing from a hadronisation vertex getting equal attribution to each hard parton's descendants entering.

However, the final products of hadronisation form distinct clusters whose centres match the momenta of the hard partons incident to the hadronisation vertex.

Take descendants from DAG, and where there is overlap in ancestry, provide functionality to attribute particles to their nearest parent in delta R.

The interface and data representations for this need careful thought. Considerations are:

1. overlap only occurs after hadronisation
2. hadronisation vertices are clearly important, how should these be represented?
3. hard partons may have descendants incident on multiple hadronisation vertices: should this be transparent in the heritage represented using the masks?
4. when a W decays into quarks, the quarks are also part of the hard process and may produce distinct jets: should MaskGroups be used to represent the composite nature?

**Comment by jacanchaplais** [comment]
The solution I've settled on is to create a function which takes the descendant masks of all of the non-incoming hard partons, and nests them into mask groups describing the decay channels. For example if the hard process is `p p > t t~`, with `t > b W+ > b u d~` and equivalent for the `t~` case, then a `MaskGroup` of `MaskGroup`s will be created with

```python
>>> masks = gcl.select.hierarchical(graph)
>>> masks
MaskGroup(mask_arrays=['t', 't~'])
```

will be created at the top level. Subscripting to get the `t` mask will yield another `MaskGroup`, _eg._

```python
>>> masks["t"]
MaskGroup(mask_arrays=['b', 'W+', 'latent'])
```
where `t` is considered a composite of the descendants of `b` and `W+`, and `latent` refers to the particles descending from `t` which are not descended from these (usually just the duplicates of `t`, and the `b` and `W+` partons themselves - since they are not included as part of their descendants).

Similarly

```python
>>> masks["t"]["W+"]
MaskGroup(mask_arrays=['u', 'd~', 'latent'])
```

and the `u`, `d~`, and `b` are `MaskArray`s rather than `MaskGroup`s, because they are the leaves of the hard process tree.

The function has an option `use_pmu: bool = True`, which then takes any descendants with overlapping heritage, and assigns them into exclusive groups, such that they are marked as the descendants of whichever partons they are closest to in the pseudorapidity-azimuth plane. This implementation is a bit lazy, it would be better to start from the final state particles, assign those based on their proximity to the hard partons, and work recursively up to the hadronisation vertex propagating that classification. However, if the emissions after hadronisation don't create emissions at wide angles, this shouldn't be a problem, and for my purposes I only care about the final state after hadronisation. Still, it's worth checking that this is a sensible approach by visualising a few events.

**Comment by jacanchaplais** [comment]
Want to split the leaf-level filtering into functions which act on the lowest level `MaskGroup` objects. One of them can be `partition_hard()`, and another could be `filter_background()`. This could then either be applied after the `hierarchy()` tree is generated, or during when each leaf-level `MaskGroup` is complete.

Would also be nice to create an iterator over the leaves of nested `MaskGroup` trees.

**Comment by jacanchaplais** [comment]
> Want to split the leaf-level filtering into functions which act on the lowest level `MaskGroup` objects. One of them can be `partition_hard()`, and another could be `filter_background()`. This could then either be applied after the `hierarchy()` tree is generated, or during when each leaf-level `MaskGroup` is complete.
> 
> Would also be nice to create an iterator over the leaves of nested `MaskGroup` trees.

Could levy the strategy pattern to pass the `hierarchy()` function post-processing functions on the leaves, so both `partition_hard()` and `filter_background()` could be passed, but also it would allow user extensibility.

**Comment by jacanchaplais** [comment]
Almost complete, just need to modify the masks so that the hard partons in the latent masks go into their respective children masks, instead. Once that is complete, the PR can be merged, and the new version can be shipped.
